p2p/net/swarm/dial_worker_test.go

```diff
@@ -342,3 +358,206 @@ func TestDialWorkerLoopConcurrentFailureStress(t *testing.T) {
+	s1.dialRanker = quicFirstRanker
+	s2.ListenClose(quicAddr)
+	s1.Peerstore().AddAddrs(s2.LocalPeer(), []ma.Multiaddr{quicAddr, tcpAddr}, peerstore.PermanentAddrTTL)
+	reqch = make(chan dialRequest)
+	resch = make(chan dialResponse)
+	worker2 := newDialWorker(s1, s2.LocalPeer(), reqch)
+	go worker2.loop()
+	defer worker2.wg.Wait()
+
+	reqch <- dialRequest{ctx: context.Background(), resch: resch}
+	select {
+	case res := <-resch:
+		t.Fatalf("expected a delay before connecting %s", res.conn.LocalMultiaddr())
+	case <-time.After(400 * time.Millisecond):
+	}
+	select {
+	case res := <-resch:
+		require.NoError(t, res.err)
+		if _, err := res.conn.LocalMultiaddr().ValueForProtocol(ma.P_TCP); err != nil {
+			t.Fatalf("expected tcp connection %s", res.conn.LocalMultiaddr())
+		}
+	case <-time.After(1 * time.Second):
+		t.Fatal("dial didn't complete")
+	}
+	close(reqch)
+	s1.ClosePeer(s2.LocalPeer())
+	s1.peers.ClearAddrs(s2.LocalPeer())
+	s2.Listen(quicAddr)
+
+	// should dial tcp immediately if there's no quic address available
+	s1.Peerstore().AddAddrs(s2.LocalPeer(), []ma.Multiaddr{tcpAddr}, peerstore.PermanentAddrTTL)
+	reqch = make(chan dialRequest)
+	resch = make(chan dialResponse)
+	worker3 := newDialWorker(s1, s2.LocalPeer(), reqch)
+	go worker3.loop()
+	defer worker3.wg.Wait()
+
+	reqch <- dialRequest{ctx: context.Background(), resch: resch}
+	select {
+	case res := <-resch:
+		require.NoError(t, res.err)
+		if _, err := res.conn.LocalMultiaddr().ValueForProtocol(ma.P_TCP); err != nil {
+			t.Fatalf("expected tcp connection, got: %s", res.conn.LocalMultiaddr())
+		}
+	case <-time.After(500 * time.Millisecond):
+		t.Fatal("dial didn't complete")
+	}
+	close(reqch)
+	s1.ClosePeer(s2.LocalPeer())
+	s1.peers.ClearAddrs(s2.LocalPeer())
+
+	// should dial next immediately when one connection errors after timeout
+	quicFirstLargeDelayRanker := func(addrs []ma.Multiaddr) []*network.AddrDelay {
+		m := make([]*network.AddrDelay, 0)
+		for _, a := range addrs {
+			if _, err := a.ValueForProtocol(ma.P_TCP); err == nil {
+				m = append(m, &network.AddrDelay{Addr: a, Delay: 10 * time.Second})
+			} else {
+				m = append(m, &network.AddrDelay{Addr: a, Delay: 0})
+			}
+		}
+		return m
+	}
+
+	s1.dialRanker = quicFirstLargeDelayRanker
+	s2.ListenClose(quicAddr)
+	s1.Peerstore().AddAddrs(s2.LocalPeer(), []ma.Multiaddr{tcpAddr, quicAddr}, peerstore.PermanentAddrTTL)
+	reqch = make(chan dialRequest)
+	resch = make(chan dialResponse)
+	worker4 := newDialWorker(s1, s2.LocalPeer(), reqch)
+	go worker4.loop()
+	defer worker4.wg.Wait()
+
+	reqch <- dialRequest{ctx: context.Background(), resch: resch}
+	select {
+	case res := <-resch:
+		require.NoError(t, res.err)
+		if _, err := res.conn.LocalMultiaddr().ValueForProtocol(ma.P_TCP); err != nil {
+			t.Fatal("expected tcp connection")
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatal("dial didn't complete")
+	}
+	close(reqch)
+	s1.ClosePeer(s2.LocalPeer())
+	s1.peers.ClearAddrs(s2.LocalPeer())
```

Comment:
Would be great if we could format these tests as tables. I think it'll be clearer.

Reply:
made this table driven.